crates/interpreter/src/interpreter/shared_stack.rs

```diff
@@ -0,0 +1,710 @@
+        if slice.is_empty() {
+            return Ok(());
+        }
+
+        let n_words = (slice.len() + 31) / 32;
+        let new_context_len = self.len() + n_words;
+        let new_buffer_len = self.buffer().len() + n_words;
+
+        if new_context_len > STACK_LIMIT {
+            return Err(InstructionResult::StackOverflow);
+        }
+
+        // SAFETY: length checked above.
+        unsafe {
+            let dst = self
+                .buffer_mut()
+                .as_mut_ptr()
+                .add(self.buffer().len())
+                .cast::<u64>();
+            let mut i = 0;
+
+            // write full words
+            let limbs = slice.rchunks_exact(8);
+            let rem = limbs.remainder();
+            for limb in limbs {
+                *dst.add(i) = u64::from_be_bytes(limb.try_into().unwrap());
+                i += 1;
+            }
+
+            // write remainder by padding with zeros
+            if !rem.is_empty() {
+                let mut tmp = [0u8; 8];
+                tmp[8 - rem.len()..].copy_from_slice(rem);
+                *dst.add(i) = u64::from_be_bytes(tmp);
+                i += 1;
+            }
+
+            debug_assert_eq!((i + 3) / 4, n_words, "wrote beyond end of stack");
+
+            // zero out upper bytes of last word
+            let m = i % 4; // 32 / 8
+            if m != 0 {
+                dst.add(i).write_bytes(0, 4 - m);
+            }
+
+            self.context_len = new_context_len;
+            self.buffer_mut().set_len(new_buffer_len);
+        }
+
+        Ok(())
+    }
+
+    /// Set a value at given index for the stack, where the top of the
+    /// stack is at index `0`. If the index is too large,
+    /// `StackError::Underflow` is returned.
+    #[inline]
+    pub fn set(&mut self, no_from_top: usize, val: U256) -> Result<(), InstructionResult> {
+        if self.len() > no_from_top {
+            let buffer = self.buffer_mut();
+            let buf_len = buffer.len();
+            // SAFETY: `0 < no_from_top + 1` <= self.len()` <= `self.buffer.len()`.
+            // Therefore, this index is bounded between 0 and `self.buffer.len()`
+            unsafe { *buffer.get_unchecked_mut(buf_len - no_from_top - 1) = val };
+            Ok(())
+        } else {
+            Err(InstructionResult::StackUnderflow)
+        }
+    }
+
+    #[inline]
+    fn page(&self) -> &Page {
+        // SAFETY: by construction, `self.page_idx` is a valid
+        // index of `self.pages`
+        unsafe { self.pages.get_unchecked(self.page_idx) }
+    }
+
+    #[inline]
+    fn page_mut(&mut self) -> &mut Page {
+        // SAFETY: by construction, `self.page_idx` is a valid
+        // index of `self.pages`
+        unsafe { self.pages.get_unchecked_mut(self.page_idx) }
+    }
+
+    #[inline]
+    fn buffer(&self) -> &Buffer {
+        // SAFETY: by construction, `self.buffer` is a valid
+        // pointer to the buffer of the current page
+        unsafe { &*self.buffer }
```

Comment:
This would become `&mut self.page.buffer` after this change: https://github.com/bluealloy/revm/pull/816/files#r1385180054